Two-channel graph (Graph2) meter style

Explorer09 opened the pull request:
This pull request implements the "two-channel" graph style as the new "Graph2" meter mode. This is based on #714 so that the graphs will be shown with colors.

This new mode is designed for a few meters only— only `DiskIORateMeter` (<s>#1763</s>, merged) and `NetworkIOMeter` can correctly utilise this mode by showing the input and output rates as separate channels.

The feature was previously discussed in #1390. 

Below is a preview screenshot taken in macOS Sequoia.
I included both Unicode mode and `--no-unicode` mode.
<img width="512" height="100" alt="htop-screenshot-graph2-meter-mode" src="https://github.com/user-attachments/assets/2eee1a9e-7383-4dab-90e0-71006268348c" />

Explorer09 committed:
Round graph meter dynamic scale and print graph scale

Round the graph meter's dynamic scale to a power of two and print the
graph scale. For a percent graph, a "%" character is printed in place
of the scale.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Allow guest time to be hidden in "non-detailed" CPU meter

When both the options "Add guest time in CPU meter percentage" and
"Detailed CPU time" are turned off, the CPU meter used to show the
"virtual" CPU time in the bar display as a glitch. Now fix it.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

eworm-de commented:
The traditional graphical meters have a gray baseline. Does it make sense to have that here as well?

Explorer09 committed:
Introduce powerOf2Floor() and popCount8() functions

This is a prerequisite for the feature "Graph meter coloring (with
GraphData structure rework)".

powerOf2Floor() will utilize __builtin_clz() or stdc_bit_floor_ui()
(__builtin_clz() is preferred) if either is supported.

popCount8() will utilize ARM NEON instructions and x86 POPCNT
instruction if the machine supports either of them.

I am not adopting the C23 standard interface stdc_count_ones_uc() yet,
as I am not sure C libraries would implement it as fast as our version.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Graph meter coloring (with GraphData structure rework)

Rewrite the entire graph meter drawing code to support color graph
drawing in addition to the dynamic scaling (both can work together
because of the new GraphData structure design).

The colors of a graph are based on the percentages of item values of
the meter. The rounding differences of each terminal character are
addressed through the different numbers of braille dots.

Due to low resolution of the character terminal, the rasterized colors
may not look nice, but better than nothing. :)

The new GraphData structure design has two technical limitations:
* The height of a graph meter now has a limit of 8191 (terminal rows).
* If dynamic scaling is used, the "total" value or scale of a graph now
  has to align to a power of 2.

The code is designed with the anticipation that the graph height may
change at runtime. No UI or option has been implemented for that yet.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Remove unused constant defines of graph meter code

Specifically 'PIXPERROW_*' and 'GraphMeterMode_dots*' constants.
They were commented out rather than removed in the previous commit (for
ease of code reviewing). Now this commit removes the constant defines
for good.

Explorer09 committed:
Introduce "Graph2" meter style (two-channel graph)

Introduce a new meter mode (meter style) called "Graph2". The "Graph2"
name stands for "two-channel graph" or "two-quadrant graph".

This meter mode is not in the list of default supported modes. This
"Graph2" mode is intended for meters with two item values. It plots
the first item above the x-axis and the second item below the x-axis.
Items other than the first two are always hidden.

This code is written with the intention to share as much code with the
Graph meter mode as possible.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
NetworkIOMeter: add "Graph2" meter mode support

Explorer09 committed:
DiskIOMeter: add "Graph2" meter mode support

Specifically it's the DiskIORateMeter part that supports "Graph2" mode.
For the combined DiskIOMeter, the "IO rate" part can show as two-channel
graph while the "IO time" part remains a one-channel graph.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Add a special "details" value for "Graph2" meter drawing

This special case of "details" value would print as ':' in the ASCII
display mode. It is used in "Graph2" meter drawing only.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
netbsd & openbsd: Simplify CPU meter value assignment code

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Revert "Fix CPU virtualization bar color and help text in non-detailed mode"

This reverts commit 47aeb0a1f282fe294c6ba1e5dbebe05157b6e88c.

Explorer09 committed:
Restore "virtualized" word for non-detailed CPU meter in help screen

The time displayed is (steal+guest), not just guest CPU time.

Regression from 3d8fa0b926a463006773534411df91be2b9d68ed

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Use CPU_METER_STEAL for 'virtualized' CPU time in non-detailed mode

Linux and PCP platform code used to write the 'virtualized'
(steal+guest) CPU time into the CPU_METER_IRQ item when the "detailed
CPU time" option is off, which would result in a wrong color painted for
virtual CPU time in bar mode. Write the 'virtualized' CPU time to
CPU_METER_STEAL instead, which is more appropriate for the job.

Also update the "virtualized" CPU meter item in the help screen: (1) The
color is now CPU_STEAL for consistency. (2) In monochrome mode, two
dummy items are displayed before the "virtualized" word so that the
bar meter symbol mapping is correct.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 commented on linux/Platform.c:
```diff
@@ -370,8 +370,14 @@ double Platform_setCPUValues(Meter* this, unsigned int cpu) {
       v[CPU_METER_KERNEL] = cpuData->systemAllPeriod / total * 100.0;
-      v[CPU_METER_IRQ] = (cpuData->stealPeriod + cpuData->guestPeriod) / total * 100.0;
-      this->curItems = 4;
+      this->curItems = 3;
+
+      v[CPU_METER_IRQ] = 0.0; // Accounted in 'kernel'
+      v[CPU_METER_SOFTIRQ] = 0.0; // Accounted in 'kernel'
+      v[CPU_METER_STEAL] = (cpuData->stealPeriod + cpuData->guestPeriod) / total * 100.0;
+      if (settings->accountGuestInCPUMeter) {
+         this->curItems = 6;
+      }
```
The issue of `CPU_METER_STEAL`  is moot as this pull request is rebased on the new version of pull request #1922.

Explorer09 commented on netbsd/Platform.c:
```diff
@@ -267,16 +267,10 @@ double Platform_setCPUValues(Meter* this, int cpu) {
+      this->curItems = 4;
    } else {
       v[CPU_METER_KERNEL] = cpuData->sysAllPeriod / total * 100.0;
-      v[CPU_METER_IRQ] = 0.0; // No steal nor guest on NetBSD
-      this->curItems = 4;
+      this->curItems = 3;
    }
    totalPercent = v[CPU_METER_NICE] + v[CPU_METER_NORMAL] + v[CPU_METER_KERNEL] + v[CPU_METER_IRQ];
```
The issue of `CPU_METER_IRQ` is moot as this pull request is rebased on the new version of pull request https://github.com/htop-dev/htop/pull/1922.

Explorer09 commented on openbsd/Platform.c:
```diff
@@ -218,16 +218,10 @@ double Platform_setCPUValues(Meter* this, unsigned int cpu) {
+      this->curItems = 4;
    } else {
       v[CPU_METER_KERNEL] = cpuData->sysAllPeriod / total * 100.0;
-      v[CPU_METER_IRQ] = 0.0; // No steal nor guest on OpenBSD
-      this->curItems = 4;
+      this->curItems = 3;
    }
    totalPercent = v[CPU_METER_NICE] + v[CPU_METER_NORMAL] + v[CPU_METER_KERNEL] + v[CPU_METER_IRQ];
```
The issue of `CPU_METER_IRQ` is moot as this pull request is rebased on the new version of pull request https://github.com/htop-dev/htop/pull/1922.

Explorer09 commented on pcp/Platform.c:
```diff
@@ -574,9 +574,15 @@ static double Platform_setOneCPUValues(Meter* this, const Settings* settings, pm
       v[CPU_METER_KERNEL] = values[CPU_SYSTEM_ALL_PERIOD].ull / total * 100.0;
+      this->curItems = 3;
+
+      v[CPU_METER_IRQ] = 0.0;
+      v[CPU_METER_SOFTIRQ] = 0.0;
       value = values[CPU_STEAL_PERIOD].ull + values[CPU_GUEST_PERIOD].ull;
-      v[CPU_METER_IRQ] = value / total * 100.0;
-      this->curItems = 4;
+      v[CPU_METER_STEAL] = value / total * 100.0;
+      if (settings->accountGuestInCPUMeter) {
+         this->curItems = 6;
+      }
```
The issue of `CPU_METER_STEAL` is moot as this pull request is rebased on the new version of pull request https://github.com/htop-dev/htop/pull/1922.
